presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.sketch.kll;
+
+import com.facebook.presto.common.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.sketch.theta.ThetaSketchStateFactory;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateMetadata;
+import org.apache.datasketches.kll.KllItemsSketch;
```

Comment:
There are some pitfalls (mainly memory-wise) we use KLL in presto_cpp, not sure if they are applicable to this Java implementation or not:
1. The sketch is initialized with a fairly large pre-allocated array depending on the error bound.  This would cause large memory usage when the number of groups are high.
2. The serialized form is not very compact and can occupy more memory than it needs.  This is ok for the partial aggregation side but would cause memory issue on the final aggregation side.

Reply:
Thanks for bringing up these points! Was there anything you did to mitigate them?

re: 1. I haven't checked the code, but didn't consider the performance with a high number of groups since the main target use case for introducing the function is to provide table and partition-level statistics for the optimizer. If you have any thoughts on optimization I would happily discuss.

re: 2: I haven't measured the serialized size but according to the charts on the Apache Datasketches website, a "standard" k value of 200 produces sketches in the range of 2-4k bytes serialized with floats after consuming 1M+ items. Let's generously assume that with doubles, we would consume double the space, so about 8k.

![kll size](https://datasketches.apache.org/docs/img/kll/kll200-vs-ds128-size.png).

Do you know which parameters effect the number of partial aggregations sent to the final agg? I assumed it's the size of the cluster but, I could be wrong. Even if coalescing 1k KLL sketches in one node this should only consume 8k * 1000 = ~8MB of memory. I feel that shouldn't pose any issue. But it really depends on how many sketches are sent to the final agg step.

Reply:
@Yuhta what sort of memory issues were observed?

Reply:
For 1 it's easy to fix in the sketch itself, just need to make the sketch grows the bottom layer when the number of elements it keeps is small.  You may need to submit PR to `datasketches` project to fix it though, or copy the code over.

For 2 the size of serialization also depends on how the input data looks like.  The worst case comparing to Q/T-digest comes from a lot of identical inputs.  In this case Q-digest does not increase the memory usage but KLL does, because it keeps samples of inputs instead of distribution.  This multiplied by the number of groups and the fan-in factor of final aggregation sometimes blows it up.  One mitigation we did is to compact the sketch before the serialization, this does not resolve the problem completely but we have been good since then on production data.

Both issues can be fixed later though, I just mention it here so we are not surprised when these issues showing up.